File: creational/factory_method/factory_method.rb
class Product
    def operation
        raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
    end
end

class ConcreteProductA < Product
    def operation
        '{Result of the ConcreteProductA}'
    end
end

class ConcreteProductB < Product
    def operation
        '{Result of the ConcreteProductB}'
    end
end

class Creator
    def factory_method
        raise NotImplementedError, "#{self.class} has not implemented method '#{__method__}'"
    end

    def some_operation
        product = factory_method
        result = product.operation
    end
end

class ConcreteCreatorA < Creator
    def factory_method
        ConcreteProductA.new
    end
end

class ConcreteCreatorB < Creator
    def factory_method
        ConcreteProductB.new
    end
end

def client_code(creator)
    print "Client: I'm not aware of the creator's class, but it still works.\n"\
          "#{creator.some_operation}"
end

client_code(ConcreteCreatorA.new)

File: creational/factory_method/factory_method.py
from __future__ import annotations
from abc import ABC, abstractmethod

class Product(ABC):
    
    @abstractmethod
    def operation(self) -> str:
        pass

class ConcreteProductA(Product):

    def operation(self) -> str:
        return "Result of the ConcreteProductA"

class ConcreteProductB(Product):

    def operation(self) -> str:
        return "Result of the ConcreteProductB"

class Creator(ABC):

    @abstractmethod
    def factory_method(self):
        pass

    def some_operation(self) -> str:
        product = self.factory_method()
        result = product.operation()
        return result

class ConcreteCreatorA(Creator):

    def factory_method(self) -> ConcreteProductA:
        return ConcreteProductA()

class ConcreteCreatorB(Creator):

    def factory_method(self) -> ConcreteProductA:
        return ConcreteProductB()

def client_code(creator: Creator) -> None:
    print(f"Client: I'm not aware of the creator's class, but it still works.\n"
          f"{creator.some_operation()}", end="")

if __name__ == "__main__":
    client_code(ConcreteCreatorA())
